blog introducing java controller framework (#16135)
kubernetes/website: content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md

---
layout: blog

> title: "Cloud Native Java: Develop kubernetes controller application"

Reviewer: ```suggestion
title: "Develop a Kubernetes controller in Java"
```

Author: revised the title

date: 2019-08-28
slug: Cloud-Native-Java-Develop-Kubernetes-Controller-Application
---

**Authors:** Min Kim (Ant Financial), Tony Ado (Ant Financial)

> The [Kubernetes offical Java SDK](https://github.com/kubernetes-client/java) project 

Reviewer: ```suggestion
The official [Kubernetes Java SDK](https://github.com/kubernetes-client/java) project 
```

Author: updated

recently released their latest work on providing the Java Kubernetes developers 
a handy Kubernetes controller-builder SDK which is helpful for easily developing 
advanced workloads or systems.  

## Overall

Java is no doubt one of the most popular programming languages in the world but
it's been difficult for a period time for those non-Golang developers to build up 
their customized controller/operator due to the lack of library resources in the 
community. In the world of Golang, there're already some excellent controller 
frameworks, for example, [controller runtime](https://github.com/kubernetes-sigs/controller-runtime),

> [operator SDK](https://github.com/operator-framework/operator-sdk) and all these

Reviewer: ```suggestion
[operator SDK](https://github.com/operator-framework/operator-sdk). These
```

Author: updated

existing Golang frameworks are relying on the various utilities from the 
[Kubernetes Golang SDK](https://github.com/kubernetes/client-go) proven to 
be stable over years. Driven by the emerging need of further integration into 

> the platform of Kubernetes, we not only ported many essential toolings from Golang 

Reviewer: ```suggestion
the platform of Kubernetes, we not only ported many essential toolings from the Golang
```

Author: update

SDK into the kubernetes Java SDK including informers, work-queues, leader-elections, 
etc. but also developed a controller-builder SDK which wires up everything into 
a runnable controller without hiccups.

## Backgrounds

> Before we dig further, we have to know why we probably need Java in Kubernetes:

Reviewer: ```suggestion
Why use Java to implement Kubernetes tooling? You might pick Java for:
```

Author: copy/pasted

- __Integrating legacy enterprise Java systems__: Many companies have their legacy 
systems or frameworks written in Java in favor of stability. We are not able to 

> move everything to Golang straightly.

Reviewer: straightly? I'm not sure what you mean here.

Author: i replaced w/ "easily"

                (workQueue) -> ControllerBuilder.controllerWatchBuilder(V1Node.class, workQueue).build())
            .withReconciler(nodeReconciler) // required, set the actual reconciler
            .withName("node-printing-controller") // optional, set name for controller for logging, thread-tracing
            .withWorkerCount(4) // optional, set worker thread count
            .withReadyFunc( nodeInformer::hasSynced) // optional, only starts controller when the cache has synced up
            .build();
```

If you notice, the new Java controller framework learnt a lot from the design of 
[controller-runtime](https://github.com/kubernetes-sigs/controller-runtime) which
successfully encapsulates the complex components inside controller into several 
clean interfaces. With the help of Java Generics, we even move on a bit and simply 
the encapsulation in a better way.

As for more advanced usage, we can wrap multiple controllers into a controller-manager 
or a leader-electing controller which helps deploying in HA setup. In a word, we can 
basically find most of the equivalence implementations here from Golang SDK and
more advanced features are under active development by us.

> ## Planned for Future 

Reviewer: ```suggestion
## Future steps
```

Author: updated

The community behind the official Kubernetes Java SDK project will be focusing on
providing more useful utilities for developers who hope to program cloud native
Java applications to extend Kubernetes. If you are interested in more details, 

> please check out our repo [kubernetes-client/java](https://github.com/kubernetes-client/java).

Reviewer: ```suggestion
please look at [kubernetes-client/java](https://github.com/kubernetes-client/java) on GitHub.
```

Author: updated

> Feel free to share also your feedback with us, through Issues or Slack.

Reviewer: Will readers know what Slack is, and if so which channel in which workspace?

Author: added link to the channel
